### PR TITLE
personal message for issue

### PR DESCRIPTION
**Pull request by QvodSoldier**
#### What type of this PR
/kind feature

#### What this PR does / why we need it:
add personal message for issue subscriber

#### Which issue(s) this PR fixes:
- [Erda Cloud Issue Link](https://terminus-org.app.terminus.io/erda/workBench/projects/387/issues/all?id=57395&issueFilter__urlQuery=eyJzdGF0ZUJlbG9uZ3MiOlsiT1BFTiIsIldPUktJTkciLCJXT05URklYIiwiUkVPUEVOIiwiUkVTT0xWRUQiXSwiYXNzaWduZWVJRHMiOlsiMTAwMDQzNiJdfQ%3D%3D&issueViewGroup__urlQuery=eyJ2YWx1ZSI6ImthbmJhbiIsImNoaWxkcmVuVmFsdWUiOnsia2FuYmFuIjoiZGVhZGxpbmUifX0%3D&iterationID=429&type=REQUIREMENT)

#### Specified Reviewers:

/assign @Effet

### CODE REVIEW

File: apistructs/event.go [hunk 1]
```diff
@@ -167,9 +171,60 @@ type IssueEvent struct {
...
+				ie.Content.StreamParams.CommentTime, ie.Content.Content)
+		}
+	}
+
+	params["issueMboxContent"] = strings.Replace(content, "\\n", "\n", -1)
+	params["issueEmailContent"] = strings.Replace(strings.Replace(content, "\\n", "</br>", -1), "\n", "</br>", -1)
+
+	// todo 换行符还有问题
+	logrus.Infof("issueMboxContent is: %s", params["issueMboxContent"])
+	logrus.Infof("issueEmailContent is: %s", params["issueEmailContent"])
```
Comment (Effet) [inline apistructs/event.go:227]: should be debug level

**Review by Effet — CHANGES_REQUESTED** [review]

File: modules/core-services/services/mbox/mbox.go [hunk 2]
```diff
@@ -51,7 +54,56 @@ func WithBundle(bdl *bundle.Bundle) Option {
...
+	switch mbox.Status {
+	case apistructs.MBoxUnReadStatus:
+		mbox.UnreadCount++
+	case apistructs.MBoxReadStatus:
+		mbox.UnreadCount = 1
+		mbox.Status = apistructs.MBoxUnReadStatus
+	}
+	mbox.Content = createReq.Content
+	mbox.Title = createReq.Title
+	mbox.CreatedAt = time.Now()
```
[the remark this answers is not in the document]
Comment (QvodSoldier) [reply modules/core-services/services/mbox/mbox.go:106]: The newly created duplicate mbox needs to be folded and displayed, and sorted according to the latest creation time.

File: modules/dop/dao/issue.go [hunk 3]
```diff
@@ -501,3 +501,14 @@ func (client *DBClient) GetIssueBugByRange(req apistructs.IssuesStageRequest) ([
 	}
 	return issue, total, nil
 }
+
+// GetReceiversByIssueID get receivers of issue event
+func (client *DBClient) GetReceiversByIssueID(issueID int64) []string {
+	var receivers []string
+	issue, _ := client.GetIssue(issueID)
+	subscribers, _ := client.GetIssueSubscribersSliceByIssueID(issueID)
```
[the remark this answers is not in the document]
Comment (QvodSoldier) [reply modules/dop/dao/issue.go:509]: done

File: modules/dop/services/issuestream/issue_stream.go [hunk 4]
```diff
@@ -84,33 +84,50 @@ func (s *IssueStream) Create(req *apistructs.IssueStreamCreateRequest) (int64, e
...
-		}()
 	}
 
+	// only send issue update event when creat comment, other event like
+	// change assignee will send event in updateIssue
+	// TODO: delete send event in updateIssue
+	// if req.StreamType == apistructs.ISTComment {
+	go func() {
+		content, _ := GetDefaultContent(is.StreamType, is.StreamParams)
+		logrus.Infof("old content is: %s", content)
```
[the remark this answers is not in the document]
Comment (QvodSoldier) [reply modules/dop/services/issuestream/issue_stream.go:95]: done

File: modules/dop/services/issue/issue.go [hunk 5]
```diff
@@ -1414,3 +1393,98 @@ func (svc *Issue) Unsubscribe(id int64, identityInfo apistructs.IdentityInfo) er
...
+	if err := svc.db.BatchDeleteIssueSubscribers(req.IssueID, needDeletedSubscribers); err != nil {
+		return err
+	}
+
+	var subscribes []dao.IssueSubscriber
+	issueID := int64(req.IssueID)
+	for k := range subscriberMap {
+		subscribes = append(subscribes, dao.IssueSubscriber{IssueID: issueID, UserID: k})
+	}
+	if err := svc.db.BatchCreateIssueSubscribers(subscribes); err != nil {
```
[the remark this answers is not in the document]
Comment (QvodSoldier) [reply modules/dop/services/issue/issue.go:1451]: done

File: modules/dop/services/issue/issue.go [hunk 6]
```diff
@@ -1414,3 +1393,98 @@ func (svc *Issue) Unsubscribe(id int64, identityInfo apistructs.IdentityInfo) er
...
+	for _, v := range oldSubscribers {
+		_, exist := subscriberMap[v.UserID]
+		if exist {
+			delete(subscriberMap, v.UserID)
+		} else {
+			needDeletedSubscribers = append(needDeletedSubscribers, v.UserID)
+		}
+	}
+
+	if err := svc.db.BatchDeleteIssueSubscribers(req.IssueID, needDeletedSubscribers); err != nil {
```
[the remark this answers is not in the document]
Comment (QvodSoldier) [reply modules/dop/services/issue/issue.go:1442]: done

File: modules/dop/services/issue/issue.go [hunk 7]
```diff
@@ -972,32 +968,15 @@ func (svc *Issue) CreateStream(updateReq apistructs.IssueUpdateRequest, streamFi
 			continue
 		}
 
+		logrus.Infof("is: why????")
```
[the remark this answers is not in the document]
Comment (QvodSoldier) [reply modules/dop/services/issue/issue.go:971]: hhhhh, why. done

File: modules/core-services/dao/mbox.go [hunk 8]
```diff
@@ -89,7 +84,7 @@ func (client *DBClient) QueryMBox(request *apistructs.QueryMBoxRequest) (*apistr
 	if request.Status != "" {
 		query = query.Where("status =? ", request.Status)
 	}
-	query = query.Order("created_at desc").Select("id,title,status,label,created_at,read_at")
+	query = query.Order("created_at desc").Select("id,title,status,label,created_at,read_at,deduplicate_id,unread_count")
```
[the remark this answers is not in the document]
Comment (QvodSoldier) [reply modules/core-services/dao/mbox.go:87]: The old code is like this, so I didn’t change it

File: modules/core-services/model/mbox.go [hunk 9]
```diff
@@ -28,6 +28,9 @@ type MBox struct {
 	Status  apistructs.MBoxStatus //read|unread
 	OrgID   int64
 	ReadAt  *time.Time
+	// The UnreadCount not null only when DeduplicateID isn't null
```
[the remark this answers is not in the document]
Comment (QvodSoldier) [reply modules/core-services/model/mbox.go:31]: done

File: apistructs/event.go [hunk 10]
```diff
@@ -167,9 +171,60 @@ type IssueEvent struct {
...
+	Params       map[string]string `json:"params"`
+}
+
+// GenEventParams generate params of issue event
+func (ie *IssueEvent) GenEventParams(locale, uiPublicURL string) map[string]string {
+	params := ie.Content.Params
+	params["issueType"] = ie.Content.IssueType.String()
+	params["issueTitle"] = ie.Content.Title
+	content := ie.Content.Content
+	logrus.Infof("content is: %s", content)
```
[the remark this answers is not in the document]
Comment (QvodSoldier) [reply apistructs/event.go:190]: done

File: apistructs/event.go (hunk 1 above)
Comment (QvodSoldier) [reply apistructs/event.go:227]: done

**Comment by QvodSoldier** [comment]
My network is broken, i will push in at night.

**Comment by QvodSoldier** [comment]
i already push it

**Review by Effet — CHANGES_REQUESTED** [review]
there are remain not resolved

File: modules/dop/endpoints/issue_stream.go [hunk 11]
```diff
@@ -53,6 +54,9 @@ func (e *Endpoints) CreateCommentIssueStream(ctx context.Context, r *http.Reques
 	var istParam apistructs.ISTParam
 	if createReq.Type == apistructs.ISTComment {
 		istParam.Comment = createReq.Content
+		istParam.CommentTime = time.Now().Format("2006-01-02 15:04:05")
+		user, _ := e.bdl.GetCurrentUser(identityInfo.UserID)
+		istParam.UserName = user.Name
```
[the remark this answers is not in the document]
Comment (QvodSoldier) [reply modules/dop/endpoints/issue_stream.go]: done

File: modules/dop/endpoints/issue_callback.go [hunk 12]
```diff
@@ -102,3 +111,35 @@ func (e *Endpoints) processIssueEvent(req apistructs.IssueEvent) error {
...
+	var emailAddrs []string
+	users, err := e.bdl.ListUsers(apistructs.UserListRequest{Plaintext: true, UserIDs: req.Content.Receivers})
+	if err != nil {
+		return err
+	}
+	for _, v := range users.Users {
+		emailAddrs = append(emailAddrs, v.Email)
+	}
+
+	logrus.Infof("params of issue event is: %v", params)
```
[the remark this answers is not in the document]
Comment (QvodSoldier) [reply modules/dop/endpoints/issue_callback.go:135]: done

**Review by Effet — APPROVED** [review]

**Pull request merged**